File: examples/F90/bufr_get_string_array.f90
!
!Copyright 2005-2016 ECMWF.
!
! This software is licensed under the terms of the Apache Licence Version 2.0
!which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
!
! In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
! virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
!
!
! FORTRAN 90 Implementation: bufr_get_string_array
!
! Description: how to get an array of strings from a BUFR message.
 
program bufr_get_string_array
  use eccodes
  implicit none
  integer            :: ifile
  integer            :: iret,i,n
  integer            :: ibufr
  integer            :: strsize
  integer, parameter  :: max_strsize = 20
  character(len=max_strsize) , dimension(:),allocatable   :: stationOrSiteName 

  call codes_open_file(ifile,'../../data/bufr/pgps_110.bufr','r')

  call codes_bufr_new_from_file(ifile,ibufr,iret)

  ! unpack the data values
  call codes_set(ibufr,'unpack',1)

  ! get the width of the strings which is the same for all of them
  call codes_get(ibufr,'stationOrSiteName->width',strsize)

  ! the width is given in bits
  strsize=strsize/8

  ! max_strsize has to be set to a value >= to the size of the strings that we are getting 
  ! back from the call to codes_get_string_array
  if (strsize > max_strsize) then
    print *,'stationOrSiteName array dimension is ',max_strsize,' and should be ',strsize
    call exit(1)
  end if

  ! allocating the array of strings to be passed to codes_get_string_array is mandatory
  call codes_get_size(ibufr,'stationOrSiteName',n)
  allocate(stationOrSiteName(n))

  ! passing an array of strings stationOrSiteName which must be allocated beforehand 
  call codes_get_string_array(ibufr,'stationOrSiteName',stationOrSiteName)
  do i=1,n
    write(*,'(A)')trim(stationOrSiteName(i))
  end do

  !remember to deallocate 
  deallocate(stationOrSiteName)

  ! release memory associated with bufr handle
  ! ibufr won't be accessible after this
  call codes_release(ibufr)

  ! close file  
  call codes_close_file(ifile)

end program bufr_get_string_array

File: examples/C/bufr_get_string_array.c
/*
 * Copyright 2005-2018 ECMWF.
 *
 * This software is licensed under the terms of the Apache Licence Version 2.0
 * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
 * virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
 */

/*
 * C Implementation: bufr_get_string_array
 *
 * Description: how to get an array of strings from a BUFR message.
 *
 */

#include "eccodes.h"

int main(int argc,char* argv[])
{
    codes_handle* h=NULL;

    size_t i=0, size = 0;
    size_t len=0;
    int err=0, cnt=0;
    char** strArray = NULL; /* array of strings */
    const char* infile = "../../data/bufr/pgps_110.bufr";

    FILE* in = fopen(infile,"r");
    if (!in) {
        printf("ERROR: unable to open file %s\n", infile);
        return 1;
    }

    /* loop over the messages in the bufr file */
    while ((h = codes_handle_new_from_file(NULL, in, PRODUCT_BUFR, &err)) != NULL || err != CODES_SUCCESS)
    {
        if (h == NULL) {
            printf("Error: unable to create handle for message %d\n",cnt);
            cnt++;
            continue;
        }

        /* we need to instruct ecCodes to expand the descriptors
          i.e. unpack the data values */
        CODES_CHECK(codes_set_long(h, "unpack", 1),0);

        /* get the size and allocate memory */
        CODES_CHECK(codes_get_size(h, "stationOrSiteName", &len), 0);
        strArray = (char**)malloc(len * sizeof(char*));
        if (!strArray) {
            printf("Error: Could not allocate memory\n");
            return 1;
        }
        CODES_CHECK(codes_get_string_array(h, "stationOrSiteName", strArray, &size), 0);
        for(i=0; i<size; ++i) {
            printf("%s\n", strArray[i]);
        }
        printf("stationOrSiteName string array size = %lu\n", (unsigned long)size);

        /* free allocated arrays */
        for(i=0; i<size; ++i) free( strArray[i] );
        free(strArray);

        /* delete handle */
        codes_handle_delete(h);

        cnt++;
    }

    fclose(in);
    return 0;
}
